### DISCUSSION TITLE
Write to a TcpStream while waiting for it to become writable from another task/thread

### DISCUSSION THREAD

**Discussion by OverShifted** [tokio-rs/tokio#5453] [Q&A] [answered]
I have a program that tries to "pipe" two streams together (a `TcpStream` and a bidirectional stream of data to/from libcurl).

I spawn two tasks to do this. One to forward TcpStream's data to libcurl and another to forward data the other way. But my program ends up in a deadlock.

In my first task, I wait for my `TcpStream` (actually it is wrapped in `Arc<tokio::sync::RwLock<TcpStream>>`) to become readable:
```rust
stream.read().await.readable().await.unwrap();
```

In my second task, I write to the `TcpStream`:
```rust
let mut stream = stream.write().await;
println!("Locked!");
stream.write_all(&buffer[..num_bytes]).await.unwrap();
```

But `Locked!` never gets printed because `stream` is already locked via `stream.read()` inside another task. This means I can't write to my `TcpStream` because I'm waiting for it to become readable from another task.

Are there any "clean" solutions available? I don't consider using `UnsafeCell`s to be neat.
I probably have a similar problem with my other "stream" as well.

**Comment by Darksonn** [comment, marked as answer]
Wrapping an IO resource in a mutex or rwlock is almost wrong. You are deadlocking because the reader and the writer are blocking each other. You should split your IO resource instead. For the specific case of `TcpStream`, you can use the `TcpStream::into_split` method to get a read and write half that can be used independently.

**Reply by OverShifted** [reply]
Thanks! My tokio-related issue is solved. (But I still have the problem with libcurl)
